diffrax/_solver/kl.py

```diff
@@ -0,0 +1,291 @@
+import operator
+from typing import Optional
+
+import equinox as eqx
+import jax.tree_util as jtu
+import lineax as lx
+from jax import numpy as jnp
+from jaxtyping import Array, PyTree
+
+from .._custom_types import (
+    Args,
+    BoolScalarLike,
+    Control,
+    DenseInfo,
+    RealScalarLike,
+    VF,
+    Y,
+)
+from .._heuristics import is_sde
+from .._solution import RESULTS
+from .._term import (
+    AbstractTerm,
+    ControlTerm,
+    MultiTerm,
+    ODETerm,
+)
+from .base import (
+    _SolverState,
+    AbstractSolver,
+    AbstractWrappedSolver,
+)
+
+
+class KLState(eqx.Module):
+    """
+    The state of the SDE and the KL divergence.
+    """
+
+    y: Y
+    kl_metric: Array
+
+
+def _compute_kl_integral(
+    drift_term1: ODETerm,
+    drift_term2: ODETerm,
+    diffusion_term: ControlTerm,
+    t0: RealScalarLike,
+    y0: Y,
+    args: Args,
+    linear_solver: lx.AbstractLinearSolver,
+) -> KLState:
+    """
+    Compute the KL divergence.
+    """
+    drift1 = drift_term1.vf(t0, y0, args)
+    drift2 = drift_term2.vf(t0, y0, args)
+    drift = jtu.tree_map(operator.sub, drift1, drift2)
+
+    diffusion = diffusion_term.vf(t0, y0, args)  # assumes same diffusion
+
+    divergences = lx.linear_solve(diffusion, drift, solver=linear_solver).value
```

Comment:
This is assuming that `diffusion` is a linear operator.
Probably the best thing to do is just an `isinstance` check and raise an explicit error if not? (Or convert to `MatrixLinearOperator`?)

Reply:
Changed